### ISSUE TITLE
Lockfile is not respected when a newer version of the dependency exists in the tree

### ISSUE THREAD

**Issue by neongreen** [pnpm/pnpm#7480]
### Verify latest release

- [X] I verified that the issue exists in the latest pnpm release

### pnpm version

8.13.1

### Which area(s) of pnpm are affected? (leave empty if unsure)

Lockfile

### Reproduction steps

```bash
mkdir repro && cd repro
pnpm init
pnpm add eslint-import-resolver-node@0.3.9
pnpm add resolve@1.20.0
```

At this point, if you try `pnpm i`, you will get "Lockfile is up to date, resolution step is skipped".

Next, relax the constraint on `resolve` from `1.20.0` to `^1.20.0`.

My reasoning is: if the lockfile was compatible with package.json before, it is still compatible. Therefore the lockfile shouldn't change. However, after running `pnpm i` again, I get:

```
dependencies:
- resolve 1.20.0
+ resolve 1.22.8
```

### Describe the Bug

Here is the output of `pnpm why resolve`.

```
repro@1.0.0 /private/tmp/repro

dependencies:
eslint-import-resolver-node 0.3.9
└── resolve 1.22.8
resolve 1.20.0
```

It seems that if pnpm sees the opportunity to upgrade to a version of `resolve` that satisfies both constraints, it goes for it despite the lockfile.

### Expected Behavior

I would

* either expect the top-level `resolve` to stay at 1.20.0, since relaxing a dependency constraint can never break a lockfile;

* or expect the documentation for https://pnpm.io/npmrc#prefer-frozen-lockfile to not say "available pnpm-lock.yaml satisfies the package.json dependencies directive" (what should it say though?),

  > When set to true and the available pnpm-lock.yaml satisfies the package.json dependencies directive, a headless installation is performed. A headless installation skips all dependency resolution as it does not need to modify the lockfile.

### Which Node.js version are you using?

21.5.0

### Which operating systems have you used?

- [X] macOS
- [ ] Windows
- [ ] Linux

**Comment by KSXGitHub** [comment]
pnpm decides to change the lockfile because it detects that the requirement from your `package.json` does not match the requirement in the lockfile.

To workaround this, you can either:
1. Manually edit `importers` or `dependencies` and `devDependencies` in the lockfile.
2. Run `pnpm add resolve@^1.20.0` (with a caret) or `pnpm add resolve`.

Regarding the desired behavior, I am not sure if pnpm should continue to ignore the relaxing of range. If I relax the version range, I would expect pnpm to update my dependencies to be latest of that range.

**Comment by neongreen** [comment]
I’ll check the workarounds later, thanks!

re/ this:

> I would expect pnpm to update my dependencies to be latest of that range.

Perhaps we use lockfiles for different purposes? Mine is both a) “no surprises during deploy” and b) “no version bumps ever unless I explicitly ask for them”. I suppose you might just have (a) in mind.

**Comment by KSXGitHub** [comment]
> Perhaps we use lockfiles for different purposes? Mine is both a) “no surprises during deploy” and b) “no version bumps ever unless I explicitly ask for them”.

I only use lockfile for purpose a as well as reproducibility. If I don't ever change the specification, the dependencies will remain the same. For purpose b, I think relaxing the specification is pretty much "asking for them".

**Comment by KSXGitHub** [comment]
Actually, this is a bug. It doesn't change with only `resolve`. What changes is the transitive dependency of `eslint-import-resolver-node`.

**Comment by KSXGitHub** [comment]
I double checked again and noticed that my `pnpm why` gives a different result than yours:

```text
dependencies:
eslint-import-resolver-node 0.3.9
└── resolve 1.22.8
resolve 1.22.8
```

_(both `resolve` are `1.22.8`)_

Can you try again with pnpm v1.14.0?

**Comment by neongreen** [comment]
Hm, I get 1.22.8 even with pnpm 8.13.1 now. (And with pnpm 8.14 as well.) Not sure what happened back then — maybe I copied the wrong output somehow? (I remember trying to choose the second dependency to reproduce the issue.)

**Comment by KSXGitHub** [comment]
If you can reproduce the case where `resolve` versions are inconsistent, then we'd have a bug.

Anyway, do you have any `.npmrc` or configuration then?

**Comment by neongreen** [comment]
Even without inconsistent resolve versions --

* If we just have `resolve: 1.20.0`, then after updating package.json to say `resolve: ^1.20.0`, `pnpm i` doesn't do anything.
* On the other hand, if we have both `resolve` and `eslint-import-resolver-node`, then after updating package.json to say `resolve: ^1.20.0`, `pnpm i` upgrades resolve.

If this is the intended behavior, then I don't understand the semantics of `pnpm i` with prefer-frozen-lockfile (which is supposed to be on by default).

> Anyway, do you have any .npmrc or configuration then?

No configuration.

**Comment by KSXGitHub** [comment]
@neongreen Good point, pnpm should either change in both cases or not change in both cases.

@zkochan Which one should be the correct behavior?

**Comment by zkochan** [comment]
I don't have objections to change pnpm's behaviour in this case. If the existing resolved version satisfies the new range, we can update the range in the lockfile and consider it up-to-date.

**Comment by KSXGitHub** [comment]
@neongreen I did a bit more digging and found that pnpm only upgrades `resolve` because `eslint-import-resolver-node` requires the latest version of `resolve`, and pnpm seeks to minimize duplication and bundle size. If we remove `eslint-import-resolver-node`, pnpm would not upgrade `resolve`. I think this behavior is what most people desire, and I personally like it too. If you don't want pnpm to optimize the dependency tree, you should keep it explicitly pinned. Do you agree?

**Comment by neongreen** [comment]
I agree that it's also reasonable behavior, but then I still think the documentation should be amended.

* ✅ prefer-frozen-lockfile is set to true? -- yes, it's the default.
* ✅ available lockfile satisfies the dependencies directive? -- IMO yes, the dependencies directive is strictly more lenient than it was.
* Yet the lockfile changes.

Perhaps the docs for `prefer-frozen-lockfile` can say that the lockfile can still be modified in the interest of minimizing dependencies and bundle size, even if the conditions above hold.

Alternatively, the docs can clarify what "satisfies" means (all version range specifiers in the lockfile are exactly the same as in the package.json? idk).

---

One more point is that the current behavior seems a bit inconsistent with `--frozen-lockfile`. Presumably `--frozen-lockfile` will *never* modify the lockfile? But then imagine the following situation:

1. There are two different versions of dependency `sub` in the lockfile.
2. Both `--frozen-lockfile` and `prefer-frozen-lockfile` are fine with it.
3. Tomorrow, a new version of `sub` is published that happens to satisfy both dependency ranges.

Is `--frozen-lockfile` suddenly going to start failing? (seems bad)

Or are `--frozen-lockfile` and `prefer-frozen-lockfile` going to diverge? (`pnpm i --frozen-lockfile` succeeds but `pnpm i` still wants to update the lockfile) -- this should be documented IMO.

**Comment by zkochan** [comment]
This behavior is by design, established in https://github.com/pnpm/pnpm/issues/13778 and https://github.com/pnpm/pnpm/pull/13779 (commit 996297c32b).

When an importer's constraint is widened or relaxed (e.g. from an exact version to a caret range), both dependency resolution and the fast lockfile updater intentionally reuse the highest version already locked in the tree that satisfies the new range. This deduplicates dependencies across the graph and avoids leaving behind unpruned duplicates that full resolution would otherwise collapse. This behavior is covered by dedicated regression tests in `pnpm11/installing/deps-installer/test/install/lockedVersionReuse.ts` (`a widened range the locked version still satisfies moves to the higher locked version`).
